go/tools/builders/embed.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2017 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// embed generates a .go file from the contents of a list of data files. It is
+// invoked by go_embed_data as an action.
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"text/template"
+	"unicode/utf8"
+)
+
+const headerTpl = `// Generated by go_embed_data for {{.Label}}. DO NOT EDIT.
+
+package {{.Package}}
+
+{{if .Multi}}
+var {{.Var}} = map[string]{{.Type}}{
+{{- range $i, $f := .Sources}}
+	{{$.Key $f}}: {{$.Var}}_{{$i}},
+{{- end}}
+}
+{{end}}
+
+`
+
+func main() {
+	log.SetPrefix("embed: ")
+	log.SetFlags(0) // don't print timestamps
+	if err := run(os.Args); err != nil {
+		log.Fatal(err)
+	}
+}
+
+type configuration struct {
+	Label, Package, Var string
+	Multi               bool
+	Sources             []string
+	out, workspace      string
+	flatten, strData    bool
+}
+
+func (c *configuration) Type() string {
+	if c.strData {
+		return "string"
+	} else {
+		return "[]byte"
+	}
+}
+
+func (c *configuration) Key(filename string) string {
+	workspacePrefix := "external/" + c.workspace + "/"
+	key := filepath.FromSlash(strings.TrimPrefix(filename, workspacePrefix))
+	if c.flatten {
+		key = path.Base(filename)
+	}
+	return strconv.Quote(key)
+}
+
+func run(args []string) error {
+	c := newConfiguration(args)
+
+	f, err := os.Create(c.out)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	w := bufio.NewWriter(f)
+	defer w.Flush()
+
+	tpl := template.Must(template.New("embed").Parse(headerTpl))
```

Comment:
don't use Must in a method.   Either make this a file-level var using Must, or process the error that comes back from Parse